examples/postgres/advanced-blog-cli/src/pagination.rs

```diff
@@ -58,16 +81,22 @@ impl<T: Query> Query for Paginated<T> {
 impl<T> RunQueryDsl<PgConnection> for Paginated<T> {}
 
 impl<T> QueryFragment<Pg> for Paginated<T>
-where
-    T: QueryFragment<Pg>,
+    where
+        T: QueryFragment<Pg>,
 {
     fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, Pg>) -> QueryResult<()> {
-        out.push_sql("SELECT *, COUNT(*) OVER () FROM (");
-        self.query.walk_ast(out.reborrow())?;
-        out.push_sql(") t LIMIT ");
-        out.push_bind_param::<BigInt, _>(&self.per_page)?;
-        out.push_sql(" OFFSET ");
-        out.push_bind_param::<BigInt, _>(&self.offset)?;
+        if self.page.is_some() && self.some_per_page.is_some() {
```

Comment:
This `QueryFragment` implementation is incompatible with the declared prepared statement caching. We need to call [`AstPass::unsafe_to_cache_prepared`](https://docs.diesel.rs/master/diesel/query_builder/struct.AstPass.html#method.unsafe_to_cache_prepared) here as otherwise diesel might reuse an existing prepared statement for the "other" variant.

Reply:
@weiznich  Could you please provide some examples on how to use this [AstPass::unsafe_to_cache_prepared](https://docs.diesel.rs/master/diesel/query_builder/struct.AstPass.html#method.unsafe_to_cache_prepared)?

Reply:
Do you mean the additional query is not correct to write like this?

<img width="1369" alt="image" src="https://user-images.githubusercontent.com/3773928/200590524-60a85d4b-7138-44e5-b5c9-a50a1680f353.png">

Reply:
> Could you please provide some examples on how to use this AstPass::unsafe_to_cache_prepared?

You literally only need to call this method inside your `QueryFragment::walk_ast` method. So an example on how to use this would be literally just `out.unsafe_to_cache_prepared()`.


> Do you mean the additional query is not correct to write like this?

The problem here is that the following problem can happen:

1. You call your paginate setup with `Some(_)` as `page`. 
2. Diesel creates a prepared statement for the query generated by the `if` branch
3. You call your paginate setup with `None` as `page`
4. Diesel just sees that this is the same type based on the `QueryId` impl and skips building the query string at all as it just uses the type id to identify the already created prepared statement.

To prevent that from happening you can use `AstPass::unsafe_to_cache_prepared` to completely opt out the caching of prepared statements for this query, or you could provide a custom `QueryId` impl that sets [`QueryId::HAS_STATIC_QUERY_ID`](https://docs.diesel.rs/master/diesel/query_builder/trait.QueryId.html#associatedconstant.HAS_STATIC_QUERY_ID) to `false`, which forces the query builder to run every time the query is executed and which uses the query string as cache key.

Reply:
@weiznich Thanks for teaching!